[TRTLLM-8831][feat] Enable early exit with overlap scheduler

Funatiq opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Exposed `willCompleteNextIteration` method for improved request state querying.
  * Enhanced streaming state transitions with support for intermediate completion states.
  * Added per-request control over last-token exclusion in generation logits.

* **Bug Fixes**
  * Fixed overlap scheduler inadvertently increasing token counts in certain execution paths.

* **Tests**
  * Improved per-sequence output validation and logits verification in test suites.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up

Alternative (faster) way using CodeRabbit AI:

**[JIRA ticket/NVBugs ID/GitHub issue/None] @coderabbitai title**

NOTE: "@coderabbitai title" will be replaced by the title generated by CodeRabbit AI, that includes the "[type]" and title.
For more info, see /.coderabbit.yaml.

-->

## Description

Usually with overlap scheduler we schedule one iteration ahead. The idea here is that when a request reaches its `max_tokens` we can avoid the last extra iteration because we know that the request will finish. This also leads to a final iteration of the execution loop where no requests are scheduled (`scheduled_batch` is empty) but we still have to process the results from the previous iteration (`previous_batch` exists).

- Update `PyExecutor` to set `GENERATION_TO_COMPLETE` state for requests that will complete next iteration.
- Update `MicroBatchScheduler` bindings to skip scheduling after `GENERATION_TO_COMPLETE` state.
- Fix `_executor_loop_overlap` to finish previous batch if current batch is empty.
- Move iter_counter handling from `ModelEngine` to `PyExecutor` to account for empty batches.
- Fix logits output length with early exit.
- Fix call counts with early exit in tests.

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.


- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

Funatiq committed:
[None][feat] Enable early exit with overlap scheduler

- Update MicroBatchScheduler bindings to skip scheduling after GENERATION_TO_COMPLETE state.
- Update PyExecutor to set GENERATION_TO_COMPLETE state for requests that will complete next iteration.
- Fix _executor_loop_overlap to finish previous batch if current batch is empty.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #22295](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22295/) [ run ] triggered by Bot. Commit: `e5ab4a8`

Funatiq committed:
[fix] Skip logits and additional outputs handling in extra iteration for overlap scheduler

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
[fix] Generation logits length for overlap scheduler early exit

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
Improve test_return_logits

- Add loop over sequences in test_generate_with_return_logits and test_generate_async_with_return_logits.
- Add assertion on sequence length in test_generate_with_return_logits and test_generate_async_with_return_logits.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
[fix] Fix test_llm_api_connector

- Reduce call_counts when using the overlap scheduler.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #22769](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22769/) [ run ] completed with state `SUCCESS`. Commit: `73a01b0` 
[/LLM/main/L0_MergeRequest_PR pipeline #17169](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/17169/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #22800](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22800/) [ run ] triggered by Bot. Commit: `f0d70e8`

Funatiq commented:
/bot run

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #22925](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22925/) [ run ] triggered by Bot. Commit: `18c5f1f`

tensorrt-cicd commented:
[PR_Github #22925](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22925/) [ run ] completed with state `ABORTED`. Commit: `18c5f1f` 
[/LLM/main/L0_MergeRequest_PR pipeline #17290](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/17290/) completed with status: 'FAILURE'

Funatiq commented:
/bot run

Funatiq committed:
Refactor test_llm_perf_metrics to use context manager for LLM instance

- Updated the test to utilize a context manager for the LLM instance, improving resource management.
- Maintained existing assertions to validate performance metrics.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
[refactor] Move iter_counter handling to PyExecutor

- Moved iter_counter in PyExecutor to ensure consistency in tracking iterations.
- This allows tracking of iteration where scheduled requests are empty.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #23386](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/23386/) [ run ] completed with state `SUCCESS`. Commit: `e5bf38d` 
[/LLM/main/L0_MergeRequest_PR pipeline #17617](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/17617/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #23422](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/23422/) [ run ] triggered by Bot. Commit: `792fe92`

tensorrt-cicd commented:
[PR_Github #23504](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/23504/) [ run ] triggered by Bot. Commit: `cc046e0`

tensorrt-cicd commented:
[PR_Github #23525](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/23525/) [ run ] triggered by Bot. Commit: `ff515ce`

Funatiq commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #23539](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/23539/) [ run ] triggered by Bot. Commit: `ff515ce`

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #23732](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/23732/) [ run ] triggered by Bot. Commit: `a3a3c9f`

tensorrt-cicd commented:
[PR_Github #23732](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/23732/) [ run ] completed with state `SUCCESS`. Commit: `a3a3c9f` 
[/LLM/main/L0_MergeRequest_PR pipeline #17862](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/17862/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #23752](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/23752/) [ run ] triggered by Bot. Commit: `a3a3c9f`

tensorrt-cicd commented:
[PR_Github #23752](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/23752/) [ run ] completed with state `SUCCESS`. Commit: `a3a3c9f` 
[/LLM/main/L0_MergeRequest_PR pipeline #17878](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/17878/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #23831](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/23831/) [ run ] completed with state `SUCCESS`. Commit: `a3a3c9f` 
[/LLM/main/L0_MergeRequest_PR pipeline #17942](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/17942/) completed with status: 'FAILURE'

Funatiq commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #24497](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/24497/) [ run ] triggered by Bot. Commit: `844ac50`

tensorrt-cicd commented:
[PR_Github #24497](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/24497/) [ run ] completed with state `SUCCESS`. Commit: `844ac50` 
[/LLM/main/L0_MergeRequest_PR pipeline #18491](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/18491/) completed with status: 'FAILURE'

Funatiq commented:
/bot run --disable-fail-fast

achartier commented on tests/unittest/llmapi/test_llm_pytorch.py:
```diff
@@ -182,28 +182,30 @@ def test_llm_reward_model():
-    assert perf_metrics.first_iter is not None
-    assert perf_metrics.iter - perf_metrics.first_iter == sampling_params.max_tokens - 1
-    assert perf_metrics.last_iter == perf_metrics.iter
+    with LLM(model=llama_model_path,
```
Looks like unrelated refactoring?

Funatiq committed:
fixup! [None][feat] Enable early exit with overlap scheduler

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
fixup! [None][feat] Enable early exit with overlap scheduler

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #24677](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/24677/) [ run ] completed with state `SUCCESS`. Commit: `6348945` 
[/LLM/main/L0_MergeRequest_PR pipeline #18633](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/18633/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #24693](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/24693/) [ run ] completed with state `SUCCESS`. Commit: `6348945` 
[/LLM/main/L0_MergeRequest_PR pipeline #18647](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/18647/) completed with status: 'FAILURE'

Funatiq commented on tensorrt_llm/_torch/pyexecutor/py_executor.py:
```diff
@@ -1394,17 +1400,18 @@ def _executor_loop_overlap(self):
+                                                    use_previous_draft_tokens):
+                    target_inputs = self._handle_speculative_decoding(
+                        scheduled_batch, previous_tensors,
+                        previous_tensors_device)
```
Thanks for confirming, I added the change.

tensorrt-cicd commented:
[PR_Github #24763](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/24763/) [ run ] completed with state `SUCCESS`. Commit: `6348945` 
[/LLM/main/L0_MergeRequest_PR pipeline #18680](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/18680/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/18680/test-results/rerun_report.html) for details.
